### PR TITLE
minimal-bootstrap: point bootstrap sources to tarballs.nixos.org

### PR DESCRIPTION
**Pull request by emilytrau**
###### Description of changes

The purpose of this PR is to move the initial source tarball to an official location as discussed in https://github.com/NixOS/nixpkgs/pull/227914#discussion_r1180637857. To facilitate I have exposed the packaging steps as the `minimal-bootstrap-sources`  package.

Related #227914

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/os-specific/linux/minimal-bootstrap/stage0-posix/bootstrap-sources.nix
Comment (06kellyjac) [inline pkgs/os-specific/linux/minimal-bootstrap/stage0-posix/bootstrap-sources.nix:29]: is replacing this with a `tarballs.nixos.org` address to be a follow up PR after `minimal-bootstrap-sources` is in all-packages?

File: pkgs/os-specific/linux/minimal-bootstrap/stage0-posix/make-bootstrap-sources.nix
Comment (06kellyjac) [inline pkgs/os-specific/linux/minimal-bootstrap/stage0-posix/make-bootstrap-sources.nix:33]: nice. :+1: 

I was confused why the sha256 was changing for a second

File: pkgs/os-specific/linux/minimal-bootstrap/stage0-posix/bootstrap-sources.nix
Comment (emilytrau) [reply pkgs/os-specific/linux/minimal-bootstrap/stage0-posix/bootstrap-sources.nix:29]: It's a placeholder to be replaced before this PR is merged :>

**Comment by lovesegfault** [comment]
Hi @emilytrau!

Thrilled to see this work progress :)

Could you take a look at the format used here to request tarball uploads? https://github.com/NixOS/nixpkgs/pull/183487

If you could make a comment in that shape, it'd help me a lot in uploading this quickly

Thanks!

**Comment by emilytrau** [comment]
@lovesegfault thanks so much!

**Comment by lovesegfault** [comment]
(Might take a couple of days, I'm travelling and didn't bring my upload keys)

**Comment by amjoseph-nixpkgs** [comment]
(I am back and close to conquering my backlog)

Quick glance LGTM; will do a closer review this weekend if this hasn't been merged by then.

**Comment by Ericson2314** [comment]
Is this undrafted because the tarballs are uploaded now?

**Review by ghost — CHANGES_REQUESTED** [review]
This has to be done in two separately-merged PRs, with a Hydra build occurring after the first but before the second.  See [below](https://github.com/NixOS/nixpkgs/pull/232576#discussion_r1230389070).  Yeah, it's annoying, I know, I went through it four or five times.

**Comment by amjoseph-nixpkgs** [comment]
I have to say it is kind of a bummer that we still end up embedding a "magic hash" of some Hydra build.  I was kinda hoping the minimal-bootstrap would let us avoid doing that.

Of course, it's less-bad since that Hydra build contains no binaries, but still...

Can't you use `sharutils` to make a shar-archive and use the `nix` binary's embedded sandbox shell to unpack it?  (this trick was @alyssais's idea)

**Comment by amjoseph-nixpkgs** [comment]
> it is kind of a bummer that we still end up embedding a "magic hash" of some Hydra build

Can't we use `builtins.fetchGit`? (note that this is totally different from `fetchgit`)?

Like this: https://github.com/NixOS/nixpkgs/pull/237863

I'm testing it now but it seems to like using 100% cpu on just one core with no console output, so I guess I'll just leave it like that and check on it in the morning...

**Comment by emilytrau** [comment]
@amjoseph-nixpkgs iirc `builtins.fetchGit` isn't available in restricted evaluation environments even with a hash declared.

Also I suppose now that I think of it again, the purpose of uploading this source tarball to `tarballs.nixos.org` is really just to establish the process in which others can update it with `stage0-posix` releases, as opposed to having it on my personal GitHub repo. As it's all source code it wouldn't differ in terms of security than fetching a tarball for any other package from any 3rd party source, and the "clean room" benefits of building in hydra maybe isn't integral?

Maybe persuading upstream to publish release tarballs in `nar` or `shar` formats would also solve the issue of maintenance.

**Comment by amjoseph-nixpkgs** [comment]
> @amjoseph-nixpkgs iirc `builtins.fetchGit` isn't available in restricted evaluation environments even with a hash declared.

Actually it *is* available!

It does check the git repo url against `allowed-uris`, so the Hydra administrators would need to add `https://github.com/oriansj/stage0-posix` to that list.  This works for me:

```
nix build -f '<nixpkgs>' minimal-bootstrap.stage0-posix.src \
  --system i686-linux \
  --option restrict-eval true \
  --option allowed-uris https://github.com/oriansj/stage0-posix
```

> the purpose of uploading this source tarball to tarballs.nixos.org is really just to establish the process in which others can update it with stage0-posix releases

Okay, but any process involving `tarballs.nixos.org` is super painful.  If this one git repo was the *only* thing you needed `tarballs.nixos.org` for, finding a way to avoid that would be a pretty big win.

**Comment by emilytrau** [comment]
@amjoseph-nixpkgs another suggestion might be to move the tarball hosting repo to `nix-community`. The process of generating a `nar` from a `stage0-posix` release is almost automated and could just need someone to open a PR bumping the version whenever needed. It would definitely be easier than going through this `tarballs.nixos.org` process every time.

**Comment by amjoseph-nixpkgs** [comment]
Ah yes, now I remember why my scheme is doomed to fail:

It isn't documented in the manual, but `builder = "builtins:fetchurl"` is the "cheat code" that lets you skip the restricted-mode `allowed-uris` check.  If you use `builtins.fetchurl` or `builtins.fetchGit` or `builtins.fetchTarball` you don't get to skip that check.

There is no `"builtins:fetchgit"` or `"builtins:fetchtarball"`.

In fact, there is no "cheat-code" that lets you access any of the `libarchive` unpacking functionality except for `xz`-decompression.

That became a problem for NixOS's channels, so another "cheat-code" was added for that use case: [`builder = "builtins:unpack-channel"`](https://github.com/NixOS/nix/blob/2e47e539538e4c87849d675939d20e91f65885a6/src/nix-channel/unpack-channel.nix#L4).  But [it barfs](https://github.com/NixOS/nix/blob/2e47e539538e4c87849d675939d20e91f65885a6/src/libstore/builtins/unpack-channel.cc#L23) if the tarball contains more than one file.

So yeah, we have to put a `nar` somewhere.

Anyways, the [split-it-into-two-PRs](https://github.com/NixOS/nixpkgs/pull/232576/#discussion_r1230389070) part of my review still stands; if you could please address that I'll approve this, we can merge it, and start waiting for Hydra to do the build.

**Comment by amjoseph-nixpkgs** [comment]
> @amjoseph-nixpkgs another suggestion might be to move the tarball hosting repo to `nix-community`.

Or have somebody run a streaming HTTP GET tarball-to-narball converter and point `./pkgs/build-support/fetchurl/boot.nix` at that.

```bash
curl -o stage0-posix.nar \
  http://nixos.org/tar2nar/https://github.com/oriansj/stage0-posix/archive/3189b5f325b7ef8b88e3edec7c1cde4fce73c76c.tar
```

**Comment by Ericson2314** [comment]
I am not knowledgeable about this Hydra policy tarball stuff, and @amjoseph-nixpkgs is, so I defer to him.

**Comment by Ericson2314** [comment]
#238357 does make sense to me though. Does it make sense to you @emilytrau as a replacement?

**Comment by emilytrau** [comment]
#238357 makes a lot of sense to me as well :)

**Pull request closed without merging**
